Hazel/src/Hazel/ImGui/ImGuiConsole.cpp

```diff
@@ -0,0 +1,249 @@
+#include "hzpch.h"
+#include "ImGuiConsole.h"
+
+#include "imgui.h"
+#include "examples/imgui_impl_glfw.h"
+#include "examples/imgui_impl_opengl3.h"
+
+namespace Hazel {
+
+	float ImGuiConsole::s_DisplayScale = 1.0f;
+	uint16_t ImGuiConsole::s_MessageBufferCapacity = 200;
+	uint16_t ImGuiConsole::s_MessageBufferSize = 0;
+	uint16_t ImGuiConsole::s_MessageBufferBegin = 0;
+	ImGuiConsole::Message::Level ImGuiConsole::s_MessageBufferRenderFilter = ImGuiConsole::Message::Level::Trace;
+	std::vector<Ref<ImGuiConsole::Message>> ImGuiConsole::s_MessageBuffer(s_MessageBufferCapacity);
+	bool ImGuiConsole::s_AllowScrollingToBottom = true;
+	bool ImGuiConsole::s_RequestScrollToBottom = false;
+
+	void ImGuiConsole::AddMessage(Ref<Message> message)
+	{
+		if (message->m_Level == Message::Level::Invalid)
+			return;
+
+		*(s_MessageBuffer.begin() + s_MessageBufferBegin) = message;
```

Comment:
This overriding behaviour seems very suspect, you should be able to just have a vector and push/emplace back, if you want to pre-reserve space you can use vector.reserve(), doing this is just going to keep additional memory in-use until its overwritten which just increase memory fragmentation.

Reply:
oops, I see its an attempt at a circular buffer, in which case you can still just do 
```
s_MessageBuffer[s_MessageBufferBegin++] = message;
if (s_MessageBufferBegin == s_MessageBufferCapacity)
  s_MessageBufferBegin = 0;
```